### PR TITLE
Enable DataParallel for DeepLift/LayerDeepLift/NeuronDeepLift and its SHAP variants

### PR DESCRIPTION
**Pull request by NarineK**
This PR enables DataParallel models for DeepLift/LayerDeepLift/NeuronDeepLift and its SHAP variants.
This required to pass inputs and baselines as a tuple and concatenate those in the main module's forward_pre_hook. The output is then being chunked in the forward_hook of the same module.

In addition to that I also needed to extend `compute_layer_gradients_and_eval ` and `_forward_layer_eval_with_neuron_grads` and enable a custom function that in this case chunks the inputs of the selected `layer` and the outputs of the selected `neuron` in the layer and neuron variants.

I also enabled cross_max output for the maxpool functions.

### CODE REVIEW

File: captum/attr/_core/deep_lift.py [hunk 1]
```diff
@@ -245,31 +238,33 @@ def attribute(
...
+            self.model, (inputs, baselines), target, input_base_additional_args
         )
-        # remove forward hook set for baselines
-        for forward_handles_ref in self.forward_handles_refs:
-            forward_handles_ref.remove()
 
-        self.model.apply(self._register_hooks)
         gradients = self.gradient_func(
-            self.model,
+            wrapped_forward_func,
```
[the remark this answers is not in the document]
Comment (NarineK) [reply captum/attr/_core/deep_lift.py]: I was thinking using `additional_forward_args` and it looks like it is making things complicated - 
`run_forward` unpacks all arguments and in the pre-hook I need a tuple of inputs and baselines that do not need to be unpacked. I have tried to wrap them in extra tuple but that doesn't work quite nicely for the inputs. If there are arbitrary number of tensors in the inputs that makes things complicated.

File: captum/attr/_core/layer/layer_deep_lift.py [hunk 2]
```diff
@@ -36,16 +38,15 @@ def __init__(self, model, layer):
                           Currently, it is assumed that the inputs or the outputs
                           of the layer, depending on which one is used for
                           attribution, can only be a single tensor.
+            device_ids (list(int)): Device ID list, necessary only if forward_func
+                          applies a DataParallel model. This allows reconstruction of
+                          intermediate outputs from batched results across devices.
+                          If forward_func is given as the DataParallel model itself,
+                          then it is not necessary to provide this argument.
```
[the remark this answers is not in the document]
Comment (NarineK) [reply captum/attr/_core/layer/layer_deep_lift.py]: yeah, I guess we can remove it since it is only used in a combination with a DataParallel.

File: tests/attr/test_data_parallel.py [hunk 3]
```diff
@@ -491,7 +517,7 @@ def _data_parallel_test_assert(
             attr_orig = algorithm(model, target_layer)
             if alt_device_ids:
                 attr_dp = algorithm(
-                    dp_model.forward, target_layer, device_ids=self._alt_device_list()
+                    dp_model, target_layer, device_ids=self._alt_device_list()
```
[the remark this answers is not in the document]
Comment (NarineK) [reply tests/attr/test_data_parallel.py]: if we remove device_ids this should get fixed

File: captum/attr/_core/layer/layer_deep_lift.py [hunk 4]
```diff
@@ -233,37 +234,56 @@ def attribute(
...
+        )
+
+        gradients, attr_inputs = compute_layer_gradients_and_eval(
+            wrapped_forward_func,
             self.layer,
             inputs,
+            device_ids=self.device_ids,
             additional_forward_args=additional_forward_args,
             attribute_to_layer_input=attribute_to_layer_input,
+            output_fn=lambda out: out.chunk(2)[0],
```
[the remark this answers is not in the document]
Comment (NarineK) [reply captum/attr/_core/layer/layer_deep_lift.py]: Good point! Thank you! I must have done another pass and clean it up more.

File: captum/attr/_core/deep_lift.py [hunk 5]
```diff
@@ -656,35 +680,48 @@ def _compute_mean_across_baselines(self, inp_bsz, base_bsz, attribution):
...
+
+
+def maxpool1d(module, inputs, outputs, grad_input, grad_output, eps=1e-10):
+    input, input_ref = inputs.chunk(2)
+    output, output_ref = outputs.chunk(2)
+
+    delta_in = input - input_ref
+    delta_in = torch.cat(2 * [delta_in])
+    delta_out_xmax = torch.max(output, output_ref)
+    delta_out = torch.cat([delta_out_xmax - output_ref, output - delta_out_xmax])
```
[the remark this answers is not in the document]
Comment (NarineK) [reply captum/attr/_core/deep_lift.py]: I'll add more description but this logic is based on the logic from original SHAP implementation.

File: captum/attr/_core/layer/layer_deep_lift.py [hunk 6]
```diff
@@ -233,37 +234,56 @@ def attribute(
...
+            additional_forward_args
+        )
+        input_base_additional_args = _expand_additional_forward_args(
+            additional_forward_args, 2, ExpansionTypes.repeat
+        )
+
+        wrapped_forward_func = self._construct_forward_func(
             self.model,
+            (inputs, baselines),
+            input_base_target,
```
[the remark this answers is not in the document]
Comment (NarineK) [reply captum/attr/_core/layer/layer_deep_lift.py]: Good point added

File: captum/attr/_core/layer/layer_deep_lift.py [hunk 7]
```diff
@@ -233,37 +234,56 @@ def attribute(
...
+            input_base_additional_args,
+        )
+
+        gradients, attr_inputs = compute_layer_gradients_and_eval(
+            wrapped_forward_func,
```
[the remark this answers is not in the document]
Comment (NarineK) [reply captum/attr/_core/layer/layer_deep_lift.py]: `_construct_forward_func ` helps us to pass the `inputs, baselines` tuple-d in a right order - if we pass it as an additional args it is becoming trick to  identify matching tensor pairs.

File: captum/attr/_core/deep_lift.py [hunk 8]
```diff
@@ -656,35 +680,48 @@ def _compute_mean_across_baselines(self, inp_bsz, base_bsz, attribution):
...
+    # updating only the first half
+    new_grad_inp[0] = grad_input_unnorm - grad_input_unnorm.sum() * 1 / n
+    return new_grad_inp
+
+
+def maxpool1d(module, inputs, outputs, grad_input, grad_output, eps=1e-10):
+    input, input_ref = inputs.chunk(2)
+    output, output_ref = outputs.chunk(2)
+
+    delta_in = input - input_ref
```
[the remark this answers is not in the document]
Comment (NarineK) [reply captum/attr/_core/deep_lift.py]: yeah, I was experimenting with it and didn't move to a generic place. Thank you!

File: captum/attr/_core/deep_lift.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (NarineK) [reply captum/attr/_core/deep_lift.py]: Yeah, it is basically taking cross maximum between delta outputs for the inputs and corresponding references. It relies on the gradients for the reference values rather than the actual inputs because they are able to describe the differences and contributions better. 

I added test cases with large reference values and made sure that in the majority of the cases it is relying on the references and that it converges properly based on the completeness property.

**Comment by NarineK** [comment]
Thank you for the thorough review, @vivekmig! Addressed most comments. Let me know what do you think.

**Comment by NarineK** [comment]
Merged with the layer-tuple changes, @vivekmig !

File: captum/attr/_utils/gradient.py [hunk 9]
```diff
@@ -372,7 +385,17 @@ def compute_layer_gradients_and_eval(
         # key_list is a list of devices in appropriate ordering for concatenation.
         # If only one key exists (standard model), key list simply has one element.
         key_list = _sort_key_list(list(saved_layer.keys()), device_ids)
-        all_outputs = _reduce_list([saved_layer[device_id] for device_id in key_list])
+
+        for device_id in key_list:
+            print("saved_layer[device_id]: ", saved_layer[device_id], is_layer_tuple)
```
Comment (NarineK) [inline captum/attr/_utils/gradient.py]: remove printout

File: captum/attr/_utils/gradient.py [hunk 10]
```diff
@@ -372,7 +385,15 @@ def compute_layer_gradients_and_eval(
         # key_list is a list of devices in appropriate ordering for concatenation.
         # If only one key exists (standard model), key list simply has one element.
         key_list = _sort_key_list(list(saved_layer.keys()), device_ids)
-        all_outputs = _reduce_list([saved_layer[device_id] for device_id in key_list])
+
+        all_outputs = _reduce_list(
+            [
+                saved_layer[device_id]
+                if output_fn is None
+                else output_fn(saved_layer[device_id])
```
[the remark this answers is not in the document]
Comment (NarineK) [reply captum/attr/_utils/gradient.py]: Since we were already iterating in the loop, I added it in the loop: https://github.com/pytorch/captum/blob/master/captum/attr/_utils/gradient.py#L179

File: captum/attr/_core/deep_lift.py [hunk 11]
```diff
@@ -729,108 +726,122 @@ def _compute_mean_across_baselines(self, inp_bsz, base_bsz, attribution):
...
+        output, output_ref = outputs.chunk(2)
+
+        delta_in = input - input_ref
+        delta_in = torch.cat(2 * [delta_in])
+        # Extracts cross maximum between the outputs of maxpool for the
+        # actual inputs and its corresponding references. In case the delta outputs
+        # for the references are larger the method relies on the references and
+        # corresponding gradients to compute the multiplies and contributions.
+        delta_out_xmax = torch.max(output, output_ref)
+        delta_out = torch.cat([delta_out_xmax - output_ref, output - delta_out_xmax])
```
[the remark this answers is not in the document]
Comment (NarineK) [reply captum/attr/_core/deep_lift.py]: Good point! 2nd option is good however it doesn't work with DataParallel because the input and the baseline will be concatenated in a different order.

File: captum/attr/_utils/gradient.py [hunk 12]
```diff
@@ -360,6 +372,7 @@ def compute_layer_gradients_and_eval(
             additional_forward_args=additional_forward_args,
             attribute_to_layer_input=attribute_to_layer_input,
             forward_hook_with_return=True,
+            output_fn=output_fn,
```
[the remark this answers is not in the document]
Comment (NarineK) [reply captum/attr/_utils/gradient.py]: I'll remove it for neuron cases. We can do it without and we can think of adding it if it is needed
